[cutlass backend] Add (limited) bmm dynamic shape support

henrylhtsang opened the pull request:
Differential Revision: D73626732

In this PR, we add support for bmm dynamic shape, provided that the batch stride is the biggest in the stride for A, B, and D. For example, for A of size `(B, M, K)`, we support stride `(M*K, K, 1)` and `(M*K, 1, M)`. With this assumption, we can infer the batch stride from existing arguments.

The reason is we don't want to add 2-3 more runtime params. The concerns are complexity and possible perf regression, though we didn't verify the latter.

We can revisit this if there is a need for that.

We also remove `B = 1` for normal mm and addmm. We tested it and didn't see perf regression. But open to revisiting this as well.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

ColinPeppler commented on torch/_inductor/kernel/mm_common.py:
```diff
@@ -288,3 +289,17 @@ def has_zero_dim(size: Sequence[_IntLike]) -> bool:
+    strides = [mat1.get_stride(), mat2.get_stride(), layout.stride]
+    for size, stride in zip(sizes, strides):
+        assert len(size) == len(stride) == 3, "Expect 3D tensors"
+        if stride[0] != math.prod(size[1:]):
```
consider using sympy_product to handle symbolic shapes, but it looks like math.prod is working here too

https://github.com/pytorch/pytorch/blob/797768cd90d0984687e15f5fe0e1a4d8bf91d71a/torch/_inductor/utils.py#L363-L364

henrylhtsang commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -118,6 +120,12 @@ def init_layout_args(self) -> None:
+        if len(X.get_size()) > 2:
+            self.add_layout_arg("B", X, "size", 0)
+            # since we assume batch stride is largest
+            self.add_layout_arg("M*K", X, "stride", 0)
+            self.add_layout_arg("K*N", W, "stride", 0)
+            self.add_layout_arg("M*N", Y, "stride", 0)
```
> My gut feeling is that the more appropriate way is to handle this M*K in the template, instead of defining it as a symbol.
> 
> It'd be something like {stride(X, dim=1) * stride(X, dim=2)} for M*K.
> 
> What's your read on this?

yeah I think it would make sense, but don't have a good way to do it. The current layout arg is too inflexible.

ColinPeppler replied:
Shouldn't {stride(X, dim=1) * stride(X, dim=2)} just resolve to {M * K}?

henrylhtsang replied:
chatted offline. We are trying the multiple layout arg idea. But its a lot of changes at the end.

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -114,10 +115,14 @@ def init_layout_args(self) -> None:
-        kdim = _normalize_idx(-1, len(X.get_size()))
         self.add_layout_arg("M", X, "size", mdim)
+        self.add_layout_arg("K", X, "size", ndim)
+        self.add_layout_arg("K", W, "size", mdim)
```
more clear if you'd use smth like `w_kdim = _normalize_idx(-1, len(X.get_size())`

henrylhtsang replied:
> more clear if you'd use smth like `w_kdim = _normalize_idx(-1, len(X.get_size())`

I don't want to define 4 more dims. If we are here, inductor should have confirmed that they are all 3D

henrylhtsang replied:
well fine

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -84,17 +85,17 @@ def find_layout_arg(
         matches = [
-            arg for arg in self.layout_args.values() if arg.matches(node, attr, dim)
+            arg
+            for args in self.layout_args.values()
+            for arg in args
```
if you want to flatten this you can do smth like

`for arg in itertools.chain.from_iterable(self.layout_args.values())`

ColinPeppler commented on torch/_inductor/codegen/cuda/cuda_kernel.py:
```diff
@@ -84,17 +85,17 @@ def find_layout_arg(
             first_match = matches[0]
+            # Verify all matches have the same underlying symbolic expressions
             if not all(
-                match.node == first_match.node
-                and match.attr == first_match.attr
-                and match.dim == first_match.dim
```
does this existing code no longer work?

henrylhtsang replied:
> does this existing code no longer work?

ok you are right, existing should still work

henrylhtsang committed:
[cutlass backend] bmm dynamic shape (#152393)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/152393

Differential Revision: D73626732

henrylhtsang commented:
@pytorchbot merge
